ops/framework.py

```diff
@@ -32,6 +32,29 @@
 from ops import charm
 from ops.storage import NoSnapshotError, SQLiteStorage
 
+if typing.TYPE_CHECKING:
+    from pathlib import Path
+
+    from ops.charm import CharmMeta
+    from ops.model import Model
+
+    try:
+        from typing import Protocol, Type
+    except ModuleNotFoundError:
+        from typing_extensions import Protocol, Type
```

Comment:
Since charmcraft includes typing-extensions in its toolchain, I think it might be better to just import these from the module. That avoids the case where the behavior in typing-extensions is different from that in typing. And it would make the code feel better to avoid the try-except in the imports.

Reply:
> Yeh I think that we should be using pyright or mypy. @PietroPasotti I'd encourage you to start by running a checker like that in a tox env like `tox -e static` - perhaps we just enable the checks 1 file at a time and move that way?

I like this strategy. I'll add the env and ignore every single file. The idea is that newly added files should pass the check.
I think enabling the checks even one file at a time will be a massive refactor, in framework.py I see over 1300 errors.
It might be more manageable if as part of an initial refactor we break off some of the larger files into modules (my personal limit is ~700 lines, I see here we've crossed that line quite a bit)